src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseCodeReviewFeedbackService.java

```diff
@@ -0,0 +1,227 @@
+
+            var feedbacks = athenaResponse.stream().filter(individualFeedbackItem -> individualFeedbackItem.filePath() != null)
+                    .filter(individualFeedbackItem -> individualFeedbackItem.description() != null).map(individualFeedbackItem -> {
+                        var feedback = new Feedback();
+                        String feedbackText;
+                        if (Objects.nonNull(individualFeedbackItem.lineStart())) {
+                            if (Objects.nonNull(individualFeedbackItem.lineEnd()) && !individualFeedbackItem.lineStart().equals(individualFeedbackItem.lineEnd())) {
+                                feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s at lines %d-%d", individualFeedbackItem.filePath(),
+                                        individualFeedbackItem.lineStart(), individualFeedbackItem.lineEnd());
+                            }
+                            else {
+                                feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s at line %d", individualFeedbackItem.filePath(),
+                                        individualFeedbackItem.lineStart());
+                            }
+                            feedback.setReference(String.format("file:%s_line:%d", individualFeedbackItem.filePath(), individualFeedbackItem.lineStart()));
+                        }
+                        else {
+                            feedbackText = String.format(NON_GRADED_FEEDBACK_SUGGESTION + "File %s", individualFeedbackItem.filePath());
+                        }
+                        feedback.setText(feedbackText);
+                        feedback.setDetailText(individualFeedbackItem.description());
+                        feedback.setHasLongFeedbackText(false);
+                        feedback.setType(FeedbackType.AUTOMATIC);
+                        feedback.setCredits(0.0);
+                        return feedback;
+                    }).toList();
+
+            automaticResult.setSuccessful(true);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+
+            this.resultService.storeFeedbackInResult(automaticResult, feedbacks, true);
+
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        catch (Exception e) {
+            log.error("Could not generate feedback", e);
+            automaticResult.setSuccessful(false);
+            automaticResult.setCompletionDate(ZonedDateTime.now());
+            this.resultRepository.save(automaticResult);
+            this.programmingMessagingService.notifyUserAboutNewResult(automaticResult, participation);
+        }
+        finally {
+            programmingExerciseParticipationService.unlockStudentRepositoryAndParticipation(participation);
```

Comment:
Could an automatic-feedback-request shortly before the deadline of an exercise lead to the repository and participation being unlocked after the deadline? For example, if the generation of the Feeback takes longer than the locking of the student repositories, and after the feedback was generated, the repository is unlocked again?

Reply:
> Could an automatic-feedback-request shortly before the deadline of an exercise lead to the repository and participation being unlocked after the deadline? For example, if the generation of the Feeback takes longer than the locking of the student repositories, and after the feedback was generated, the repository is unlocked again?

Very good point! I just tried this on ma-schwind. While the server does prevent further changes to the repository via the code editor (and returns a 403), the "Your repository is locked." indicator is gone and the editor is not in readonly mode.

Reply:
Very good point! Fixed, thanks.